teps/0059-skipping-strategies.md

```diff
@@ -0,0 +1,898 @@
+verbosity. 
+
+## Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+
+Unit and integration tests for guarded execution of `Tasks` in a `Pipeline` with different skipping strategies in 
+different combinations.
+
+## Design Evaluation
+
+<!--
+How does this proposal affect the reusability, simplicity, flexibility 
+and conformance of Tekton, as described in [design principles](https://github.com/tektoncd/community/blob/master/design-principles.md)
+-->
+
+### Reusability
+
+By unblocking the execution of dependent `Tasks` when a guarded `Task` is skipped, we enable execution to continue 
+when the guarded `Task` is either successful or skipped, making the `Pipeline` reusable for more scenarios or use 
+cases. 
+
+By cascading `WhenExpressions` or composing a set of `Tasks` as a `Pipeline` in a `Pipeline`, we reuse existing features
+to provide flexible skipping strategies. 
+
+### Simplicity
+
+By scoping the skipping strategy to `WhenExpressions` only, we provide the flexibility safely with a minimal change. 
+We also limit the interleaving of `Pipeline` graph execution paths and maintain the simplicity of the workflows.
+
+`WhenExpressions` and `Pipelines` in `Pipelines` are the bare minimum features so solve for the CI/CD use cases that 
+need skipping strategies. 
+
+### Flexibility
+
+This TEP will give users the flexibility to either guard a `Task` only or guard a `Task` and its dependent `Tasks`. 
+Moreover, it gives users the flexibility to guard some dependent `Tasks` while executing other dependent `Tasks`. 
+
+## Upgrade & Migration Strategy
+
+<!--
+Use this section to detail whether this feature needs an upgrade or
+migration strategy. This is especially useful when we modify a
+behavior or add a feature that may replace and deprecate a current one.
+-->
+
+Changing the scope of `WhenExpressions` to guard the `Task` only is backwards-incompatible, so to make the 
+transition smooth:
+- we'll provide a feature flag, `set-when-expressions-scope-to-task`, which:
+  - will default to `set-when-expressions-scope-to-task` : `false` to guard a `Task` and its dependent `Tasks`
+  - can be set to `set-when-expressions-scope-to-task` : `true` to guard a `Task` only 
+- after 9 months, we'll flip the feature flag and default to `set-when-expressions-scope-to-task` : `true`
+- after 9 more months, we'll remove the feature flag and `WhenExpressions` will be scoped to guard a `Task` only going 
+  forward
```

Comment:
hey @jerop i have a late request - im wondering how we want to handle this when we do the v1 release? the 9 months requirement really only applies to the beta types themselves, so with v1 we have a chance to flip the behavior:

- but it would be weird to have beta + v1 behave differently by default
- BUT it would be even weirder to do the v1 release with default behavior we already know we want to change

basically im hoping we can plan to release v1 with the preferred default 🙏  and wondering what your thoughts are (might make sense to add to the TEP also)

Reply:
hey @bobcatfish 👋🏾

wrote that assuming that in 9 months when we switch to guarding task by default, we'd have not yet released v1 -- but happy to adjust the timeline if we think we'll release v1 sooner 

looks that v1 is still a ways off - https://github.com/tektoncd/pipeline/issues/3548 - what do you think @vdemeester?

Reply:
updates made in https://github.com/tektoncd/community/pull/481, thanks @bobcatfish 😁